File: submissions/1146-greatest-common-divisor-of-strings/solution.java
class Solution {
    public String gcdOfStrings(String str1, String str2) {
        if(str1.length() > str2.length()) {
            return gcdHelper(str1, str2);
        } else {
            return gcdHelper(str2, str1);
        }
        
    }

    public String gcdHelper(String longStr, String shortStr) {
        
        if (shortStr.length()==0) {
            return "";
        }
        if (longStr.equals(shortStr)) {
            return longStr;
        }

        int longLength = longStr.length();
        int shortLength = shortStr.length();

        int mostRepeat = longLength / shortLength;

        String possibleCommon = shortStr.repeat(mostRepeat);
        if (longStr.equals(possibleCommon)) {
            return shortStr;
        }

        if (longStr.startsWith(possibleCommon)) {
            return gcdHelper(shortStr, longStr.substring(possibleCommon.length()));
        } else {
            return "";
        }
    }

    
}

File: submissions/1146-greatest-common-divisor-of-strings/solution.py
class Solution:
    def gcdOfStrings(self, str1: str, str2: str) -> str:
        if str1 == str2:
            return str1
        
        if len(str1) > len(str2):
            big = str1
            small = str2
        elif len(str1) < len(str2):
            big = str2
            small = str1
        else: # equal but not the same, they don't have gcd
            return ""

        while big != small:
            smallLen = len(small)
            while big.startswith(small):
                big = big[smallLen:]
            
            if len(big) == 0:
                return small
            if len(big) >= smallLen:
                return ""
            
            big, small = small, big

        return ""

        
